### PR TITLE
Remove coupling of DynamoDbLockClient from S3 storage

### PR DESCRIPTION
**Pull request by ahmedriza**
# Description
As part of the refactoring done by https://github.com/delta-io/delta-rs/pull/508, the `LockClient` used by the S3 storage implementation got coupled to a specific implementation of the `LockClient`, i.e. `DynamoDbLockClient`.  That doesn't look to be necessary and this PR addresses that by removing that coupling and let the S3 storage support any `LockClient`. 

This is useful in cases where we may not be using S3 in AWS.  For example, one could be using MinIO, Hitachi HCP etc which implement the S3 protocol and can be used just like S3.  In these cases, a DynamoDb based lock is not required and some other implementation can be used.

# Related Issue(s)

# Documentation

### CODE REVIEW

**Review by rtyler — COMMENTED** [review]
Nice little refactor! I like it, but will wait for @mosyp the original contributor of this code to approve before merge

**Review by mosyp — APPROVED** [review]
Nice one, thank you @ahmedriza

**Pull request merged**
